### PR TITLE
Cut 0.0.104

### PR DESCRIPTION
**Pull request by TheBlueMatt**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,92 @@
...
+ * `user_payment_id` has been removed from `PaymentPurpose`, and
+   `ChannelManager::create_inbound_payment{,_for_hash}` (#1180).
+ * Updated documentation for several `ChannelManager` functions to remove stale
+   references to panics which no longer occur (#1201).
+ * The `Score` and `LockableScore` objects have moved into the
+   `routing::scoring` module instead of being in the `routing` module (#1166).
+ * The `Time` parameter to `ScorerWithTime` is no longer longer exposed,
+   instead being fixed based on the `std`/`no-std` feature (#1184).
+ * `ChannelDetails::balance_msat` was added to access a channel's balance
+   without the reserve values, lining up with on-chain claim amounts (#1203).
```
Comment (jkczyz) [inline CHANGELOG.md:39]: Should this say "with the reserve value" or "without subtracting the reserve value"? I read "without" as "not including".

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,92 @@
...
+   such info (#1178, #1197).
+ * `Score::channel_penalty_msat` has additional arguments describing the
+   channel's capacity and the HTLC amount being sent over the channel (#1166).
+ * A new log level `Gossip` has been added, which is used for verbose
+   information generated during network graph sync. Enabling the
+   `max_level_trace` feature or ignoring `Gossip` log entries reduces log
+   growth during initial start up from many GiB to several MiB (#1145).
+ * The `allow_wallclock_use` feature has been removed in favor of only using
+   the `std` and `no-std` features (#1212).
+ * `NetworkGraph` can now remove channels which we haven't heard updates for in
```
Comment (jkczyz) [inline CHANGELOG.md:24]: s/which/that

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,92 @@
...
+   references to panics which no longer occur (#1201).
+ * The `Score` and `LockableScore` objects have moved into the
+   `routing::scoring` module instead of being in the `routing` module (#1166).
+ * The `Time` parameter to `ScorerWithTime` is no longer longer exposed,
+   instead being fixed based on the `std`/`no-std` feature (#1184).
+ * `ChannelDetails::balance_msat` was added to access a channel's balance
+   without the reserve values, lining up with on-chain claim amounts (#1203).
+ * An explicit `UserConfig::accept_inbound_channels` flag is now provided,
+   removing the need to set `min_funding_satoshis` to > 21 million BTC (#1173).
+ * Inbound Channels which fail to see the funding transaction confirm within
```
Comment (jkczyz) [inline CHANGELOG.md:42]: s/Channels/channels
s/which/that

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,92 @@
...
+ * `NetAddress::OnionV2` parsing was removed as version 2 onion services are no
+   longer supported in modern Tor (#1204).
+ * Generation and signing of anchor outputs is now supported in the
+   `KeysInterface`, though no support for them exists in the channel itself (#1176)
+
+## Bug Fixes
+ * Fixed a race condition in `InvoicePayer` where paths may be retried after
+   the retry count has been exceeded. In this case the
+   `Event::PaymentPathFailed::all_paths_failed` field is not a reliable payment
+   failure indicator. There is no acceptable alternative indicator (#1202).
```
Comment (jkczyz) [inline CHANGELOG.md:56]: Maybe mention `Event::PaymentFailed`?

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,92 @@
...
+ * Reduced the blocks-before-timeout we expect of outgoing HTLCs before
+   refusing to forward. This check was overly strict and resulted in refusing
+   to forward some HTLCs to a next hop which had a lower security threshold
```
Comment (jkczyz) [inline CHANGELOG.md:59]: s/which/that

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,92 @@
...
+   than us (#1119).
+ * LDK no longer attempt to update the channel fee for outbound channels when
+   we cannot afford the new fee. This could have caused force-closure by our
+   channel counterparty (#1054).
+ * Fixed several bugs which may have prevented the reliable broadcast of our
+   own channel announcements and updates (#1169).
+ * Fixed a rare bug which may have resulted in spurious route finding failures
+   when using last-hop hints and MPP with large value payments (#1168).
+ * `KeysManager::spend_spendable_outputs` no longer adds a change output which
```
Comment (jkczyz) [inline CHANGELOG.md:68]: s/which/that

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,92 @@
...
+   is below the dust threshold for non-standard change scripts (#1131).
+ * Fixed a minor memory leak when attempting to send a payment which fails due
```
Comment (jkczyz) [inline CHANGELOG.md:70]: s/which/that

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,92 @@
...
+   to an error when updating the `ChannelMonitor` (#1143).
+ * Fixed a bug where a `FeeEstimator` which returns values which are rounded to
```
Comment (jkczyz) [inline CHANGELOG.md:72]: s/which returns values which are rounded/that returns fees rounded

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,92 @@
...
+   the next sat/vbyte may result in force-closures (#1208).
+ * Handle MPP timeout HTLC error codes, instead of considering the recipient to
+   have sent an invalid error, removing them from the network graph (#1148)
+
+## Serialization Compatibility
+ * All above new Events/fields are ignored by prior clients. All above new
+   Events/fields are not present when reading objects serialized by prior
+   versions of the library.
```
Comment (jkczyz) [inline CHANGELOG.md:80]: Use either "events" or `Event`s in two places.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,92 @@
+# 0.0.104 - XXX
```
Comment (jkczyz) [inline CHANGELOG.md:1]: Update date when ready.

**Review by jkczyz — COMMENTED** [review]
Mostly obligatory grammar nits 🙂

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,92 @@
...
+   versions of the library.
+ * Payment secrets are now generated deterministically. This reduces the memory
+   footprint for inbound payments, however, newly-generated inbound payments
+   using `ChannelManager::create_inbound_payment{,_for_hash}` will not be
+   receivable using versions prior to 0.0.104.
+   `ChannelManager::create_inbound_payment{,_for_hash}_legacy` are provided for
+   backwards compatibility (#1177).
+ * `PaymentPurpose::InvoicePayment::user_payment_id` will be 0 when reading
+   objects written with 0.0.104 when read by 0.0.103 and previous (#1180).
+
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply CHANGELOG.md:89]: Yep, was waiting to make sure we get all the PRs in

**Review by valentinewallace — COMMENTED** [review]
ACK after squash

**Comment by TheBlueMatt** [comment]
Squashed, diff since original PR:

```diff
$ git diff-tree -U1 e1ffc78759d15a1589b1bedde73212ef67c52af3 ec86e2a1
diff --git a/CHANGELOG.md b/CHANGELOG.md
index a862da9c5..df1204e9b 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,2 +1,2 @@
-# 0.0.104 - XXX
+# 0.0.104 - 2021-12-17
 
@@ -23,3 +23,3 @@
    the `std` and `no-std` features (#1212).
- * `NetworkGraph` can now remove channels which we haven't heard updates for in
+ * `NetworkGraph` can now remove channels that we haven't heard updates for in
    two weeks with `NetworkGraph::remove_stale_channels{,with_time}`. The first
@@ -37,7 +37,8 @@
    instead being fixed based on the `std`/`no-std` feature (#1184).
- * `ChannelDetails::balance_msat` was added to access a channel's balance
-   without the reserve values, lining up with on-chain claim amounts (#1203).
+ * `ChannelDetails::balance_msat` was added to fetch a channel's balance
+   without subtracting the reserve values, lining up with on-chain claim amounts
+   less on-chain fees (#1203).
  * An explicit `UserConfig::accept_inbound_channels` flag is now provided,
    removing the need to set `min_funding_satoshis` to > 21 million BTC (#1173).
- * Inbound Channels which fail to see the funding transaction confirm within
+ * Inbound channels that fail to see the funding transaction confirm within
    2016 blocks are automatically force-closed with
@@ -55,7 +56,8 @@
    `Event::PaymentPathFailed::all_paths_failed` field is not a reliable payment
-   failure indicator. There is no acceptable alternative indicator (#1202).
+   failure indicator. There was no acceptable alternative indicator,
+   `Event::PaymentFailed` as been added to provide one (#1202).
  * Reduced the blocks-before-timeout we expect of outgoing HTLCs before
    refusing to forward. This check was overly strict and resulted in refusing
-   to forward some HTLCs to a next hop which had a lower security threshold
-   than us (#1119).
+   to forward som HTLCs to a next hop that had a lower security threshold than
+   us (#1119).
  * LDK no longer attempt to update the channel fee for outbound channels when
@@ -67,8 +69,8 @@
    when using last-hop hints and MPP with large value payments (#1168).
- * `KeysManager::spend_spendable_outputs` no longer adds a change output which
+ * `KeysManager::spend_spendable_outputs` no longer adds a change output that
    is below the dust threshold for non-standard change scripts (#1131).
- * Fixed a minor memory leak when attempting to send a payment which fails due
+ * Fixed a minor memory leak when attempting to send a payment that fails due
    to an error when updating the `ChannelMonitor` (#1143).
- * Fixed a bug where a `FeeEstimator` which returns values which are rounded to
-   the next sat/vbyte may result in force-closures (#1208).
+ * Fixed a bug where a `FeeEstimator` that returns values rounded to the next
+   sat/vbyte may result in force-closures (#1208).
  * Handle MPP timeout HTLC error codes, instead of considering the recipient to
@@ -77,4 +79,4 @@
 ## Serialization Compatibility
- * All above new Events/fields are ignored by prior clients. All above new
-   Events/fields are not present when reading objects serialized by prior
+ * All above new events/fields are ignored by prior clients. All above new
+   events/fields are not present when reading objects serialized by prior
    versions of the library.
@@ -89,2 +91,15 @@
 
+In total, this release features 51 files changed, 5356 insertions, 2238
+deletions in 107 commits from 9 authors, in alphabetical order:
+ * Antoine Riard
+ * Conor Okus
+ * Devrandom
+ * Duncan Dean
+ * Elias Rohrer
+ * Jeffrey Czyz
+ * Ken Sedgwick
+ * Matt Corallo
+ * Valentine Wallace
+
+
 # 0.0.103 - 2021-11-02
diff --git a/lightning-background-processor/Cargo.toml b/lightning-background-processor/Cargo.toml
index 531678325..0056393dd 100644
--- a/lightning-background-processor/Cargo.toml
+++ b/lightning-background-processor/Cargo.toml
@@ -2,3 +2,3 @@
 name = "lightning-background-processor"
-version = "0.0.103"
+version = "0.0.104"
 authors = ["Valentine Wallace <vwallace@protonmail.com>"]
@@ -13,7 +13,7 @@ edition = "2018"
 bitcoin = "0.27"
-lightning = { version = "0.0.103", path = "../lightning", features = ["std"] }
-lightning-persister = { version = "0.0.103", path = "../lightning-persister" }
+lightning = { version = "0.0.104", path = "../lightning", features = ["std"] }
+lightning-persister = { version = "0.0.104", path = "../lightning-persister" }
 
 [dev-dependencies]
-lightning = { version = "0.0.103", path = "../lightning", features = ["_test_utils"] }
-lightning-invoice = { version = "0.11.0", path = "../lightning-invoice" }
+lightning = { version = "0.0.104", path = "../lightning", features = ["_test_utils"] }
+lightning-invoice = { version = "0.12.0", path = "../lightning-invoice" }
diff --git a/lightning-block-sync/Cargo.toml b/lightning-block-sync/Cargo.toml
index e9d5f5281..ae6eb5825 100644
--- a/lightning-block-sync/Cargo.toml
+++ b/lightning-block-sync/Cargo.toml
@@ -2,3 +2,3 @@
 name = "lightning-block-sync"
-version = "0.0.103"
+version = "0.0.104"
 authors = ["Jeffrey Czyz", "Matt Corallo"]
@@ -17,3 +17,3 @@ rpc-client = [ "serde", "serde_json", "chunked_transfer" ]
 bitcoin = "0.27"
-lightning = { version = "0.0.103", path = "../lightning" }
+lightning = { version = "0.0.104", path = "../lightning" }
 tokio = { version = "1.0", features = [ "io-util", "net", "time" ], optional = true }
diff --git a/lightning-invoice/Cargo.toml b/lightning-invoice/Cargo.toml
index 010becc27..71648b3c8 100644
--- a/lightning-invoice/Cargo.toml
+++ b/lightning-invoice/Cargo.toml
@@ -3,3 +3,3 @@ name = "lightning-invoice"
 description = "Data structures to parse and serialize BOLT11 lightning invoices"
-version = "0.11.0"
+version = "0.12.0"
 authors = ["Sebastian Geisler <sgeisler@wh2.tu-dresden.de>"]
@@ -12,3 +12,3 @@ readme = "README.md"
 bech32 = "0.8"
-lightning = { version = "0.0.103", path = "../lightning" }
+lightning = { version = "0.0.104", path = "../lightning" }
 secp256k1 = { version = "0.20", features = ["recovery"] }
@@ -19,2 +19,2 @@ bitcoin_hashes = "0.10"
 hex = "0.3"
-lightning = { version = "0.0.103", path = "../lightning", features = ["_test_utils"] }
+lightning = { version = "0.0.104", path = "../lightning", features = ["_test_utils"] }
diff --git a/lightning-net-tokio/Cargo.toml b/lightning-net-tokio/Cargo.toml
index ad3dc008e..0a763f09f 100644
--- a/lightning-net-tokio/Cargo.toml
+++ b/lightning-net-tokio/Cargo.toml
@@ -2,3 +2,3 @@
 name = "lightning-net-tokio"
-version = "0.0.103"
+version = "0.0.104"
 authors = ["Matt Corallo"]
@@ -14,3 +14,3 @@ edition = "2018"
 bitcoin = "0.27"
-lightning = { version = "0.0.103", path = "../lightning" }
+lightning = { version = "0.0.104", path = "../lightning" }
 tokio = { version = "1.0", features = [ "io-util", "macros", "rt", "sync", "net", "time" ] }
diff --git a/lightning-persister/Cargo.toml b/lightning-persister/Cargo.toml
index ef8631059..7c32b6cf4 100644
--- a/lightning-persister/Cargo.toml
+++ b/lightning-persister/Cargo.toml
@@ -2,3 +2,3 @@
 name = "lightning-persister"
-version = "0.0.103"
+version = "0.0.104"
 authors = ["Valentine Wallace", "Matt Corallo"]
@@ -15,3 +15,3 @@ unstable = ["lightning/unstable"]
 bitcoin = "0.27"
-lightning = { version = "0.0.103", path = "../lightning" }
+lightning = { version = "0.0.104", path = "../lightning" }
 libc = "0.2"
@@ -22,2 +22,2 @@ winapi = { version = "0.3", features = ["winbase"] }
 [dev-dependencies]
-lightning = { version = "0.0.103", path = "../lightning", features = ["_test_utils"] }
+lightning = { version = "0.0.104", path = "../lightning", features = ["_test_utils"] }
diff --git a/lightning/Cargo.toml b/lightning/Cargo.toml
index b80e0a5b1..1234cd5c0 100644
--- a/lightning/Cargo.toml
+++ b/lightning/Cargo.toml
@@ -2,3 +2,3 @@
 name = "lightning"
-version = "0.0.103"
+version = "0.0.104"
 authors = ["Matt Corallo"]
$
```

**Review by jkczyz — DISMISSED** [review]

**Review by valentinewallace — DISMISSED** [review]

**Review by jkczyz — APPROVED** [review]

**Review by valentinewallace — APPROVED** [review]

**Pull request merged**